### PR TITLE
feat(cache): Add events and add all the missing update events.

### PR DESCRIPTION
**Pull request by deepsarda**
This pr, adds: -  

1. events which send the old object and the new object like discord.js.
2. `MESSAGE_REACTION_ADD, MESSAGE_REACTION_REMOVE, MESSAGE_REACTION_REMOVE_ALL, CHANNEL_UPDATE, MESSAGE_UPDATE, GUILD_UPDATE, GUILD_ROLE_UPDATE, GUILD_MEMBER_UPDATE, USER_UPDATE` to `setupCacheEdits.ts`.

### CODE REVIEW

**Comment by deepsarda** [comment]
Proposal: To remove breaking changes when migrating from the old cache plugin we remove bot.cache.guilds etc and make it bot.guilds.

**Comment by AwesomeStickz** [comment]
Personally I don't think we'd want to emit old object stuff and all since this was originally intended to be a way to handle cache only and nothing much, not sure about others' thoughts on this

**Comment by deepsarda** [comment]
> Personally I don't think we'd want to emit old object stuff and all since this was originally intended to be a way to handle cache only and nothing much, not sure about others' thoughts on this

Well, it's one of the major drawbacks of the previous cache system, which is leading many people to make their own cache plugin to handle it.

Logging functions of discord bots are completely broken.

**Comment by deepsarda** [comment]
But if it is not needed then ig I can remove it?

**Comment by AwesomeStickz** [comment]
> But if it is not needed then ig I can remove it?

That's just my personal opinion though, we'll have to wait for skillz to see if he wants that in it

**Comment by deepsarda** [comment]
> > But if it is not needed then ig I can remove it?
> 
> That's just my personal opinion though, we'll have to wait for skillz to see if he wants that in it

Okay!

File: events.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    bot: BotWithProxyCache<ProxyCacheTypes, Bot>,
+    oldRole: Role,
+    newRole: Role,
+  ): unknown;
+  GUILD_MEMBER_UPDATE_WITH_OLD_MEMBER(
+    bot: BotWithProxyCache<ProxyCacheTypes, Bot>,
+    oldMember: Member,
+    newMember: Member,
+  ): unknown;
+  USER_UPDATE_WITH_OLD_USER(
```
Comment (itohatweb) [inline events.ts]: Is there a reason for why those are capitalized?

Comment (itohatweb) [reply events.ts]: Also the type of those functions feel to be wrong.

**Review by itohatweb — CHANGES_REQUESTED** [review]
Why are there so many unrelated formatting changes?

**Comment by itohatweb** [comment]
> Proposal: To remove breaking changes when migrating from the old cache plugin we remove bot.cache.guilds etc and make it bot.guilds.

I do not really understand how that removes breaking changes?

**Comment by Skillz4Killz** [comment]
im a big fan of update events like this but its really hard to review this pr as itoh said so much formatting changes going on

File: events.ts (hunk 1 above)
Comment (deepsarda) [reply events.ts]: > Is there a reason for why those are capitalized?

Thats how the incoming events are named.

Can make them snakecased.

**Comment by deepsarda** [comment]
> im a big fan of update events like this but its really hard to review this pr as itoh said so much formatting changes going on

I ran `deno fmt` i think that caused all the changes.

**Comment by deepsarda** [comment]
@itohatweb @Skillz4Killz I hope all the formatting changes are gone.

And before we merge can we run `deno fmt`?

**Comment by deepsarda** [comment]
> > Proposal: To remove breaking changes when migrating from the old cache plugin we remove bot.cache.guilds etc and make it bot.guilds.
> 
> I do not really understand how that removes breaking changes?

With the current cache plugin, say we want to get a guild we do `client.guilds.get(id)`, but with this we have to do `client.cache.guilds.get(id)`.  I want to remove `.cache` and access guilds directly

**Comment by AwesomeStickz** [comment]
> > > Proposal: To remove breaking changes when migrating from the old cache plugin we remove bot.cache.guilds etc and make it bot.guilds.
> > 
> > 
> > I do not really understand how that removes breaking changes?
> 
> With the current cache plugin, say we want to get a guild we do `client.guilds.get(id)`, but with this we have to do `client.cache.guilds.get(id)`. I want to remove `.cache` and access guilds directly

So how about cache.options and other stuff? It stays there?

**Comment by deepsarda** [comment]
> > > > Proposal: To remove breaking changes when migrating from the old cache plugin we remove bot.cache.guilds etc and make it bot.guilds.
> > > 
> > > 
> > > I do not really understand how that removes breaking changes?
> > 
> > 
> > With the current cache plugin, say we want to get a guild we do `client.guilds.get(id)`, but with this we have to do `client.cache.guilds.get(id)`. I want to remove `.cache` and access guilds directly
> 
> So how about cache.options and other stuff? It stays there?

Yup, thats the idea

File: events.ts (hunk 1 above)
Comment (Skillz4Killz) [reply events.ts]: memberUpdate, guildUpdate etc....

**Review by Skillz4Killz — APPROVED** [review]

**Pull request merged**